### PR TITLE
New Jenkinsfile template (issue #385)

### LINKED ISSUE #385: Jenkins: new `load ${DMAKE_JENKINS_FILE_V2}` that only defines a `dmake()` groovy function
To cleanup the way dmake users projects effectively call dmake in Jenkins, we should introduce a new version of  `DMAKE_JENKINS_FILE` (https://github.com/Deepomatic/dmake/blob/master/dmake/templates/jenkins/Jenkinsfile) that only defines a groovy function `dmake()`, then the user is free to call `dmake` however he wants.

It should resolve #377, and help for #375.

Not sure about the job properties parameters though.. they would be somewhat bypassed...

# Use-case
Different dmake calls in different nodes (cf #375).
Example (untested):
```groovy
load "${DMAKE_JENKINS_FILE_V2}"
node {
    stage {
        parallel cpu: {
            node('cpu') {
              dmake(app='e2e-cpu')
          }
        },
       cpu: {
          node('cpu') {
            dmake(app='e2e-cpu')
        }
    }
  }
}
```

Later: integrate that as a Jenkins Pipeline Shared Library: https://jenkins.io/doc/book/pipeline/shared-libraries/

### PR DESCRIPTION
**Pull request by ibreschi**
Closes #385

### CODE REVIEW

File: dmake/templates/jenkins_V2/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    }
+}
+
+// Function that takes the parameters and creates a build dmake call
+def build(dmake_app_list) {
+
+    // Setup variables 
+    setup_pipeline_parameters()
+    def is_deploy_branch = (env.BRANCH_NAME == "release" || env.BRANCH_NAME == "master")
+    def NO_GPU = (! is_deploy_branch) ? 1 : 0
```
Comment (thomas-riccardi) [inline dmake/templates/jenkins_V2/Jenkinsfile]: deploy branch was a vulcan concept up until now, used for slack notification: it should be generalized as such:
- maybe ask for a list of branches to notify for;
- or maybe integrate that even deeper into dmake: in the dmake.yml somehow;
- or keep it outside of dmake for now: handled in a different PR

File: dmake/templates/jenkins_V2/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        if (abortedCauseDescription) {
+            currentBuild.result = 'ABORTED'
+            buildMessage = abortedCauseDescription
+        } else {
+            currentBuild.result = 'FAILURE'
+        }
+        throw e
+    } finally {
+        if (is_deploy_branch) {
+            notifyBuild("vulcan", currentBuild.result, buildMessage)
```
Comment (thomas-riccardi) [inline dmake/templates/jenkins_V2/Jenkinsfile]: see other comment: it should be generic and not be tied to vulcan (which is a private repo, vs dmake which is public)

File: dmake/templates/jenkins_V2/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    if (! dmake_app_list){
+	dmake_app_list = [params.DMAKE_APP]
+    }
+
+    def buildMessage = null   
+    try {
+        for (dmake_app in dmake_app_list){
+            dmake_call(environment, dmake_command, dmake_with_dependencies, dmake_app)
+        }
+        currentBuild.result = 'SUCCESS'
```
Comment (thomas-riccardi) [inline dmake/templates/jenkins_V2/Jenkinsfile]: how does `currentBuild` updates in multiple nodes in `parallel` behave?

File: dmake/templates/jenkins_V2/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            if (cause instanceof jenkins.model.CauseOfInterruption.UserInterruption) {
+                return cause.getShortDescription()
+            }
+        }
+    }
+}
+
+def setup_pipeline_parameters() {
+
+    // default parameters from dmake repo user
```
Comment (thomas-riccardi) [inline dmake/templates/jenkins_V2/Jenkinsfile]:
```suggestion
    // default parameters from dmake user repo
```

File: dmake/templates/jenkins_V2/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+              load 'DMakefile'
+        }
+    }
+}
+
+// Function that takes the parameters and creates a build dmake call
+def build(dmake_app_list) {
+
+    // Setup variables 
+    setup_pipeline_parameters()
```
Comment (thomas-riccardi) [inline dmake/templates/jenkins_V2/Jenkinsfile]: maybe auto call that on file load, and require one global load then one call to `build` per node?

File: dmake/templates/jenkins_V2/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+
+    // For PRs Jenkins will give the source branch name
+    def is_pr = !!env.CHANGE_BRANCH
+    def dmake_command = params.DMAKE_COMMAND
+    try {
+        dmake_command = OVERRIDE_DMAKE_COMMAND
+    } catch (e) {}
+
+    def environment = ["TOTO=1", "TOTO2=1", "ALLOW_NO_GPU=${NO_GPU}", "DMAKE_NO_GPU=${NO_GPU}", "VULCAN_MINIKUBE_CONTEXT=${env.MINIKUBE_CONTEXT}"]
```
Comment (thomas-riccardi) [inline dmake/templates/jenkins_V2/Jenkinsfile]: cleanup: remove `TOTO`

Comment (thomas-riccardi) [inline dmake/templates/jenkins_V2/Jenkinsfile]: these extra environment variables are vulcan specific: they should *not* be there: use `withEnv` around `build` in vulcan Jenkinsfile instead

File: dmake/templates/jenkins_V2/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+	
+        withEnv(environment){
+              sh "${params.CUSTOM_ENVIRONMENT} python3 \$(which dmake) ${dmake_command} ${dmake_with_dependencies} ${dmake_app}"
+              load 'DMakefile'
+        }
+    }
+}
+
+// Function that takes the parameters and creates a build dmake call
+def build(dmake_app_list) {
```
Comment (thomas-riccardi) [inline dmake/templates/jenkins_V2/Jenkinsfile]:
```suggestion
def execute(dmake_app_list) {
```

File: src/com/deepo/Dmake.groovy [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    def dmake_command = params.DMAKE_COMMAND
+    try {
+        dmake_command = OVERRIDE_DMAKE_COMMAND
+    } catch (e) {}
+
+    if (! dmake_command) {
+        dmake_command = Utils.build.isPR() ? 'test' : 'deploy'
+    }
+
+    def dmake_with_dependencies = params.DMAKE_WITH_DEPENDENCIES ? '' : '--standalone'
```
[the remark this answers is not in the document]
Comment (ibreschi) [reply src/com/deepo/Dmake.groovy]: I'm sorry I don't master enough dmake to understand what it means. It was in the jenkins template and I kept it

File: dmake/templates/jenkins_V2/Jenkinsfile (hunk 3 above)
Comment (ibreschi) [reply dmake/templates/jenkins_V2/Jenkinsfile]: For the moment it actually behaved as expected but. When one node succeeded and the other failed the global result was a fail.

**Comment by ibreschi** [comment]
Closed for inactivity

**Pull request closed without merging**